File: src/main/java/com/lannstark/lec07/Lec07Main.java
package com.lannstark.lec07;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class Lec07Main {

  public static void main(String[] args) throws IOException {
    JavaFilePrinter printer = new JavaFilePrinter();
    printer.readFile("./a.txt");
  }

  private int parseIntoOrThrow(@NotNull String str) {
    try {
      return Integer.parseInt(str);
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException(String.format("주어진 %s는 숫자가 아닙니다.", str));
    }
  }

  private Integer parseIntoOrThrow2(@NotNull String str) {
    try {
      return Integer.parseInt(str);
    } catch (NumberFormatException e) {
      return null;
    }
  }


}

File: src/main/kotlin/com/lannstark/lec07/Lec07Main.kt
package com.lannstark.lec07

import java.io.BufferedReader
import java.io.FileReader

fun main() {

}

// 코틀린에서는 try with resources 가 없고 use 를 사용한다.
fun readFile(path: String) {
    BufferedReader(FileReader(path)).use {
        println(it.readLine())
    }
}

fun parseIntOrThrow(str: String): Int {
    try {
        return str.toInt()
    } catch (e: NumberFormatException) {
        throw IllegalArgumentException("주어진 ${str}는 숫자가 아닙니다.")
    }
}

/*
    try 문도 expression이다. -> 즉 값을 만들어 내기 때문에 바로 return
 */
fun parseIntOrThrow2(str: String): Int? {
    return try {
        str.toInt()
    } catch (e: NumberFormatException) {
        null
    }
}
